File: internal/store/store.go
// Package store is the SQLite-backed persistence layer for Columbia Pages.
// HTML is stored inline in the database (pages are small text documents). Back
// up the main database together with its WAL state during a write pause.
package store

import (
	"database/sql"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"time"

	_ "modernc.org/sqlite" // pure-Go SQLite driver (no cgo), registers "sqlite"
)

// ErrNotFound is returned when a page does not exist.
var ErrNotFound = errors.New("page not found")

// Page is a stored page. For themed pages, HTML holds the body content that the
// server wraps in the house theme; for raw pages, HTML is a complete document
// served verbatim.
type Page struct {
	ID        string
	Title     string
	Slug      string
	HTML      string
	Raw       bool
	CreatedAt time.Time
	UpdatedAt time.Time
	ExpiresAt *time.Time // nil = never expires
}

// Meta is page metadata without the (potentially large) HTML body, for listings.
type Meta struct {
	ID        string
	Title     string
	Slug      string
	Raw       bool
	CreatedAt time.Time
	UpdatedAt time.Time
	ExpiresAt *time.Time
	Size      int // bytes of HTML
}

// Store wraps the database handle.
type Store struct {
	db *sql.DB
}

// Open opens (creating if needed) the SQLite database at path, runs migrations,
// and returns a ready Store. The parent directory is created if missing.
func Open(path string) (*Store, error) {
	if dir := filepath.Dir(path); dir != "" && dir != "." {
		if _, err := os.Stat(dir); errors.Is(err, os.ErrNotExist) {
			if err := os.MkdirAll(dir, 0o700); err != nil {
				return nil, fmt.Errorf("create db dir: %w", err)
			}
		} else if err != nil {
			return nil, fmt.Errorf("inspect db dir: %w", err)
		}
	}

	// WAL + a generous busy timeout keep the single-user workload contention-free.
	dsn := "file:" + path + "?_pragma=busy_timeout(5000)&_pragma=journal_mode(WAL)&_pragma=foreign_keys(on)"
	db, err := sql.Open("sqlite", dsn)
	if err != nil {
		return nil, fmt.Errorf("open db: %w", err)
	}
	// One connection sidesteps SQLite write-locking entirely; ample for one user.
	db.SetMaxOpenConns(1)

	s := &Store{db: db}
	if err := s.migrate(); err != nil {
		db.Close()
		return nil, err
	}
	if err := os.Chmod(path, 0o600); err != nil {
		db.Close()
		return nil, fmt.Errorf("secure db: %w", err)
	}
	return s, nil
}

// Close closes the underlying database.
func (s *Store) Close() error { return s.db.Close() }

func (s *Store) migrate() error {
	const schema = `
CREATE TABLE IF NOT EXISTS pages (
  id         TEXT PRIMARY KEY,
  title      TEXT NOT NULL,
  slug       TEXT NOT NULL DEFAULT '',
  html       TEXT NOT NULL,
  raw        INTEGER NOT NULL DEFAULT 0,
  created_at TEXT NOT NULL,
  updated_at TEXT NOT NULL,
  expires_at TEXT
);
CREATE INDEX IF NOT EXISTS idx_pages_created ON pages(created_at);
CREATE INDEX IF NOT EXISTS idx_pages_expires ON pages(expires_at);

CREATE TABLE IF NOT EXISTS device_authorizations (
  id                    TEXT PRIMARY KEY,
  device_code_hash      TEXT NOT NULL UNIQUE,
  device_secret_hash    TEXT NOT NULL,
  user_code_hash        TEXT NOT NULL UNIQUE,
  device_label          TEXT NOT NULL,
  scopes                 TEXT NOT NULL,
  source_key             TEXT NOT NULL,
  source_hint            TEXT NOT NULL,
  status                 TEXT NOT NULL,
  created_at             TEXT NOT NULL,
  expires_at             TEXT NOT NULL,
  approved_at            TEXT,
  denied_at              TEXT,
  last_poll_at           TEXT,
  poll_interval_seconds  INTEGER NOT NULL,
  consumed_at            TEXT
);
CREATE INDEX IF NOT EXISTS idx_device_authorizations_user_code ON device_authorizations(user_code_hash);
CREATE INDEX IF NOT EXISTS idx_device_authorizations_source ON device_authorizations(source_key, status);
CREATE INDEX IF NOT EXISTS idx_device_authorizations_expires ON device_authorizations(expires_at);

CREATE TABLE IF NOT EXISTS api_tokens (
  id              TEXT PRIMARY KEY,
  token_hash      TEXT NOT NULL UNIQUE,
  display_prefix  TEXT NOT NULL,
  device_label    TEXT NOT NULL,
  scopes          TEXT NOT NULL,
  created_at      TEXT NOT NULL,
  expires_at      TEXT NOT NULL,
  last_used_at    TEXT,
  revoked_at      TEXT
);
CREATE INDEX IF NOT EXISTS idx_api_tokens_hash ON api_tokens(token_hash);
CREATE INDEX IF NOT EXISTS idx_api_tokens_expires ON api_tokens(expires_at);

CREATE TABLE IF NOT EXISTS admin_sessions (
  id            TEXT PRIMARY KEY,
  session_hash  TEXT NOT NULL UNIQUE,
  authenticated INTEGER NOT NULL DEFAULT 0,
  created_at    TEXT NOT NULL,
  expires_at    TEXT NOT NULL
);
CREATE INDEX IF NOT EXISTS idx_admin_sessions_hash ON admin_sessions(session_hash);
CREATE INDEX IF NOT EXISTS idx_admin_sessions_expires ON admin_sessions(expires_at);`
	_, err := s.db.Exec(schema)
	if err != nil {
		return fmt.Errorf("migrate: %w", err)
	}
	return nil
}

const rfc = time.RFC3339Nano

func nullTime(t *time.Time) any {
	if t == nil {
		return nil
	}
	return t.UTC().Format(rfc)
}

func parseTime(s string) (time.Time, error) { return time.Parse(rfc, s) }

// Create inserts a new page.
func (s *Store) Create(p *Page) error {
	_, err := s.db.Exec(
		`INSERT INTO pages (id, title, slug, html, raw, created_at, updated_at, expires_at)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
		p.ID, p.Title, p.Slug, p.HTML, boolToInt(p.Raw),
		p.CreatedAt.UTC().Format(rfc), p.UpdatedAt.UTC().Format(rfc), nullTime(p.ExpiresAt),
	)
	if err != nil {
		return fmt.Errorf("create page: %w", err)
	}
	return nil
}

// Get returns the full page (including HTML) regardless of expiry. Callers that
// serve pages publicly should check ExpiresAt themselves.
func (s *Store) Get(id string) (*Page, error) {
	row := s.db.QueryRow(
		`SELECT id, title, slug, html, raw, created_at, updated_at, expires_at
		 FROM pages WHERE id = ?`, id)

	var p Page
	var raw int
	var created, updated string
	var expires sql.NullString
	switch err := row.Scan(&p.ID, &p.Title, &p.Slug, &p.HTML, &raw, &created, &updated, &expires); {
	case errors.Is(err, sql.ErrNoRows):
		return nil, ErrNotFound
	case err != nil:
		return nil, fmt.Errorf("get page: %w", err)
	}
	p.Raw = raw != 0
	if t, err := parseTime(created); err == nil {
		p.CreatedAt = t
	}
	if t, err := parseTime(updated); err == nil {
		p.UpdatedAt = t
	}
	if expires.Valid {
		if t, err := parseTime(expires.String); err == nil {
			p.ExpiresAt = &t
		}
	}
	return &p, nil
}

// Save overwrites an existing page's mutable fields. Returns ErrNotFound if the
// id does not exist.
func (s *Store) Save(p *Page) error {
	res, err := s.db.Exec(
		`UPDATE pages SET title = ?, slug = ?, html = ?, raw = ?, updated_at = ?, expires_at = ?
		 WHERE id = ?`,
		p.Title, p.Slug, p.HTML, boolToInt(p.Raw),
		p.UpdatedAt.UTC().Format(rfc), nullTime(p.ExpiresAt), p.ID,
	)
	if err != nil {
		return fmt.Errorf("save page: %w", err)
	}
	if n, _ := res.RowsAffected(); n == 0 {
		return ErrNotFound
	}
	return nil
}

// Delete removes a page. Returns ErrNotFound if it did not exist.
func (s *Store) Delete(id string) error {
	res, err := s.db.Exec(`DELETE FROM pages WHERE id = ?`, id)
	if err != nil {
		return fmt.Errorf("delete page: %w", err)
	}
	if n, _ := res.RowsAffected(); n == 0 {
		return ErrNotFound
	}
	return nil
}

// List returns page metadata, newest first, up to limit (limit <= 0 means all).
func (s *Store) List(limit int) ([]Meta, error) {
	q := `SELECT id, title, slug, raw, created_at, updated_at, expires_at, length(html)
	      FROM pages ORDER BY created_at DESC`
	args := []any{}
	if limit > 0 {
		q += " LIMIT ?"
		args = append(args, limit)
	}
	rows, err := s.db.Query(q, args...)
	if err != nil {
		return nil, fmt.Errorf("list pages: %w", err)
	}
	defer rows.Close()

	var out []Meta
	for rows.Next() {
		var m Meta
		var raw int
		var created, updated string
		var expires sql.NullString
		if err := rows.Scan(&m.ID, &m.Title, &m.Slug, &raw, &created, &updated, &expires, &m.Size); err != nil {
			return nil, fmt.Errorf("scan page: %w", err)
		}
		m.Raw = raw != 0
		if t, err := parseTime(created); err == nil {
			m.CreatedAt = t
		}
		if t, err := parseTime(updated); err == nil {
			m.UpdatedAt = t
		}
		if expires.Valid {
			if t, err := parseTime(expires.String); err == nil {
				m.ExpiresAt = &t
			}
		}
		out = append(out, m)
	}
	return out, rows.Err()
}

// DeleteExpired removes all pages whose expiry is at or before now. Returns the
// number deleted.
func (s *Store) DeleteExpired(now time.Time) (int, error) {
	res, err := s.db.Exec(
		`DELETE FROM pages WHERE expires_at IS NOT NULL AND expires_at <= ?`,
		now.UTC().Format(rfc))
	if err != nil {
		return 0, fmt.Errorf("delete expired: %w", err)
	}
	n, _ := res.RowsAffected()
	return int(n), nil
}

// DeleteExpiredAuth removes expired device grants, tokens, and admin sessions.
func (s *Store) DeleteExpiredAuth(now time.Time) (int, error) {
	cutoff := now.UTC().Format(rfc)
	tx, err := s.db.Begin()
	if err != nil {
		return 0, fmt.Errorf("begin auth cleanup: %w", err)
	}
	defer tx.Rollback()

	total := int64(0)
	for _, q := range []string{
		`DELETE FROM device_authorizations WHERE expires_at <= ?`,
		`DELETE FROM api_tokens WHERE expires_at <= ?`,
		`DELETE FROM admin_sessions WHERE expires_at <= ?`,
	} {
		res, err := tx.Exec(q, cutoff)
		if err != nil {
			return 0, fmt.Errorf("delete expired auth state: %w", err)
		}
		n, _ := res.RowsAffected()
		total += n
	}
	if err := tx.Commit(); err != nil {
		return 0, fmt.Errorf("commit auth cleanup: %w", err)
	}
	return int(total), nil
}

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}

File: ts/src/store/Store.ts
import { Context, Effect, Layer } from "effect"
import { SqlClient } from "effect/unstable/sql"
import type { SqlError } from "effect/unstable/sql/SqlError"
import { schema } from "./schema.ts"
import {
  type AdminSession,
  type APIToken,
  type DeviceAuthorization,
  GrantConsumed,
  GrantDenied,
  GrantExpired,
  GrantNotFound,
  GrantPending,
  LimitReached,
  type Meta,
  NotFound,
  type Page,
  SessionLimit,
  SlowDown,
} from "./models.ts"

const iso = (d: Date): string => d.toISOString()
const parseNull = (s: string | null | undefined): Date | null => (s == null ? null : new Date(s))
const boolToInt = (b: boolean): number => (b ? 1 : 0)

interface PageRow {
  readonly id: string
  readonly title: string
  readonly slug: string
  readonly html: string
  readonly raw: number
  readonly created_at: string
  readonly updated_at: string
  readonly expires_at: string | null
}
interface MetaRow extends Omit<PageRow, "html"> {
  readonly size: number
}
interface CountRow {
  readonly n: number
}
interface DeviceRow {
  readonly id: string
  readonly device_code_hash: string
  readonly device_secret_hash: string
  readonly user_code_hash: string
  readonly device_label: string
  readonly scopes: string
  readonly source_key: string
  readonly source_hint: string
  readonly status: string
  readonly created_at: string
  readonly expires_at: string
  readonly approved_at: string | null
  readonly denied_at: string | null
  readonly last_poll_at: string | null
  readonly poll_interval_seconds: number
  readonly consumed_at: string | null
}
interface TokenRow {
  readonly id: string
  readonly token_hash: string
  readonly display_prefix: string
  readonly device_label: string
  readonly scopes: string
  readonly created_at: string
  readonly expires_at: string
  readonly last_used_at: string | null
  readonly revoked_at: string | null
}
interface SessionRow {
  readonly id: string
  readonly session_hash: string
  readonly authenticated: number
  readonly created_at: string
  readonly expires_at: string
}

const toPage = (r: PageRow): Page => ({
  id: r.id,
  title: r.title,
  slug: r.slug,
  html: r.html,
  raw: r.raw !== 0,
  createdAt: new Date(r.created_at),
  updatedAt: new Date(r.updated_at),
  expiresAt: parseNull(r.expires_at),
})

const toMeta = (r: MetaRow): Meta => ({
  id: r.id,
  title: r.title,
  slug: r.slug,
  raw: r.raw !== 0,
  createdAt: new Date(r.created_at),
  updatedAt: new Date(r.updated_at),
  expiresAt: parseNull(r.expires_at),
  size: r.size,
})

const toDevice = (r: DeviceRow): DeviceAuthorization => ({
  id: r.id,
  deviceCodeHash: r.device_code_hash,
  deviceSecretHash: r.device_secret_hash,
  userCodeHash: r.user_code_hash,
  deviceLabel: r.device_label,
  scopes: r.scopes,
  sourceKey: r.source_key,
  sourceHint: r.source_hint,
  status: r.status,
  createdAt: new Date(r.created_at),
  expiresAt: new Date(r.expires_at),
  approvedAt: parseNull(r.approved_at),
  deniedAt: parseNull(r.denied_at),
  lastPollAt: parseNull(r.last_poll_at),
  pollIntervalSeconds: r.poll_interval_seconds,
  consumedAt: parseNull(r.consumed_at),
})

const toToken = (r: TokenRow): APIToken => ({
  id: r.id,
  tokenHash: r.token_hash,
  displayPrefix: r.display_prefix,
  deviceLabel: r.device_label,
  scopes: r.scopes,
  createdAt: new Date(r.created_at),
  expiresAt: new Date(r.expires_at),
  lastUsedAt: parseNull(r.last_used_at),
  revokedAt: parseNull(r.revoked_at),
})

const toSession = (r: SessionRow): AdminSession => ({
  id: r.id,
  sessionHash: r.session_hash,
  authenticated: r.authenticated !== 0,
  createdAt: new Date(r.created_at),
  expiresAt: new Date(r.expires_at),
})

const maxAdminSessions = 256

const make = Effect.gen(function* () {
  const sql = yield* SqlClient.SqlClient

  // WAL + a generous busy timeout keep the single-user workload contention-free.
  yield* sql.unsafe("PRAGMA journal_mode = WAL")
  yield* sql.unsafe("PRAGMA busy_timeout = 5000")
  yield* sql.unsafe("PRAGMA foreign_keys = ON")

  // Run migrations. Split on ';' since no statement embeds one.
  for (const stmt of schema.split(";")) {
    const trimmed = stmt.trim()
    if (trimmed !== "") yield* sql.unsafe(trimmed)
  }

  const changes = Effect.map(sql<CountRow>`SELECT changes() AS n`, (rows) => rows[0]?.n ?? 0)

  // --- pages ---------------------------------------------------------------

  const create = (p: Page): Effect.Effect<void, SqlError> =>
    sql`INSERT INTO pages (id, title, slug, html, raw, created_at, updated_at, expires_at)
        VALUES (${p.id}, ${p.title}, ${p.slug}, ${p.html}, ${boolToInt(p.raw)}, ${iso(p.createdAt)}, ${iso(p.updatedAt)}, ${p.expiresAt ? iso(p.expiresAt) : null})`.pipe(
      Effect.asVoid,
    )

  const get = (id: string): Effect.Effect<Page, SqlError | NotFound> =>
    Effect.flatMap(
      sql<PageRow>`SELECT id, title, slug, html, raw, created_at, updated_at, expires_at FROM pages WHERE id = ${id}`,
      (rows) => (rows[0] === undefined ? Effect.fail(new NotFound()) : Effect.succeed(toPage(rows[0]))),
    )

  const save = (p: Page): Effect.Effect<void, SqlError | NotFound> =>
    sql.withTransaction(
      Effect.gen(function* () {
        yield* sql`UPDATE pages SET title = ${p.title}, slug = ${p.slug}, html = ${p.html}, raw = ${boolToInt(p.raw)},
            updated_at = ${iso(p.updatedAt)}, expires_at = ${p.expiresAt ? iso(p.expiresAt) : null} WHERE id = ${p.id}`
        if ((yield* changes) === 0) return yield* Effect.fail(new NotFound())
      }),
    )

  const del = (id: string): Effect.Effect<void, SqlError | NotFound> =>
    sql.withTransaction(
      Effect.gen(function* () {
        yield* sql`DELETE FROM pages WHERE id = ${id}`
        if ((yield* changes) === 0) return yield* Effect.fail(new NotFound())
      }),
    )

  const list = (limit: number): Effect.Effect<ReadonlyArray<Meta>, SqlError> =>
    Effect.map(
      limit > 0
        ? sql<MetaRow>`SELECT id, title, slug, raw, created_at, updated_at, expires_at, length(html) AS size
             FROM pages ORDER BY created_at DESC LIMIT ${limit}`
        : sql<MetaRow>`SELECT id, title, slug, raw, created_at, updated_at, expires_at, length(html) AS size
             FROM pages ORDER BY created_at DESC`,
      (rows) => rows.map(toMeta),
    )

  const deleteExpired = (now: Date): Effect.Effect<number, SqlError> =>
    sql.withTransaction(
      Effect.gen(function* () {
        yield* sql`DELETE FROM pages WHERE expires_at IS NOT NULL AND expires_at <= ${iso(now)}`
        return yield* changes
      }),
    )

  const deleteExpiredAuth = (now: Date): Effect.Effect<number, SqlError> =>
    sql.withTransaction(
      Effect.gen(function* () {
        const cutoff = iso(now)
        let total = 0
        yield* sql`DELETE FROM device_authorizations WHERE expires_at <= ${cutoff}`
        total += yield* changes
        yield* sql`DELETE FROM api_tokens WHERE expires_at <= ${cutoff}`
        total += yield* changes
        yield* sql`DELETE FROM admin_sessions WHERE expires_at <= ${cutoff}`
        total += yield* changes
        return total
      }),
    )

  // --- device authorizations ----------------------------------------------

  const createDeviceAuthorization = (
    g: DeviceAuthorization,
    perSource: number,
    perInstance: number,
  ): Effect.Effect<void, SqlError | LimitReached> =>
    sql.withTransaction(
      Effect.gen(function* () {
        const created = iso(g.createdAt)
        const sourceCount = (yield* sql<CountRow>`SELECT count(*) AS n FROM device_authorizations
            WHERE source_key = ${g.sourceKey} AND status = 'pending' AND expires_at > ${created}`)[0]!.n
        const totalCount = (yield* sql<CountRow>`SELECT count(*) AS n FROM device_authorizations
            WHERE status = 'pending' AND expires_at > ${created}`)[0]!.n
        if (sourceCount >= perSource || totalCount >= perInstance) {
          return yield* Effect.fail(new LimitReached())
        }
        yield* sql`INSERT INTO device_authorizations
            (id, device_code_hash, device_secret_hash, user_code_hash, device_label, scopes, source_key, source_hint, status, created_at, expires_at, poll_interval_seconds)
            VALUES (${g.id}, ${g.deviceCodeHash}, ${g.deviceSecretHash}, ${g.userCodeHash}, ${g.deviceLabel}, ${g.scopes}, ${g.sourceKey}, ${g.sourceHint}, 'pending', ${created}, ${iso(g.expiresAt)}, ${g.pollIntervalSeconds})`
      }),
    )

  const deviceSelect = `SELECT id, device_code_hash, device_secret_hash, user_code_hash, device_label, scopes, source_key, source_hint,
        status, created_at, expires_at, approved_at, denied_at, last_poll_at, poll_interval_seconds, consumed_at FROM device_authorizations`

  const deviceAuthorizationByUserCode = (
    hash: string,
    now: Date,
  ): Effect.Effect<DeviceAuthorization, SqlError | GrantNotFound | GrantExpired> =>
    Effect.flatMap(
      sql<DeviceRow>`${sql.unsafe(deviceSelect)} WHERE user_code_hash = ${hash}`,
      (rows): Effect.Effect<DeviceAuthorization, GrantNotFound | GrantExpired> => {
        if (rows[0] === undefined) return Effect.fail(new GrantNotFound())
        const g = toDevice(rows[0])
        return g.expiresAt.getTime() > now.getTime() ? Effect.succeed(g) : Effect.fail(new GrantExpired())
      },
    )

  const deviceAuthorizationByDeviceCode = (
    codeHash: string,
    secretHash: string,
    now: Date,
  ): Effect.Effect<DeviceAuthorization, SqlError | GrantNotFound | GrantExpired> =>
    Effect.flatMap(
      sql<DeviceRow>`${sql.unsafe(deviceSelect)} WHERE device_code_hash = ${codeHash} AND device_secret_hash = ${secretHash}`,
      (rows): Effect.Effect<DeviceAuthorization, GrantNotFound | GrantExpired> => {
        if (rows[0] === undefined) return Effect.fail(new GrantNotFound())
        const g = toDevice(rows[0])
        return g.expiresAt.getTime() > now.getTime() ? Effect.succeed(g) : Effect.fail(new GrantExpired())
      },
    )

  const decideDeviceAuthorization = (
    hash: string,
    decision: "approved" | "denied",
    now: Date,
  ): Effect.Effect<void, SqlError | GrantNotFound> =>
    sql.withTransaction(
      Effect.gen(function* () {
        const column = decision === "denied" ? sql.literal("denied_at") : sql.literal("approved_at")
        yield* sql`UPDATE device_authorizations SET status = ${decision}, ${column} = ${iso(now)}
            WHERE user_code_hash = ${hash} AND status = 'pending' AND expires_at > ${iso(now)}`
        if ((yield* changes) === 0) return yield* Effect.fail(new GrantNotFound())
      }),
    )

  /**
   * pollDeviceAuthorization advances the persisted polling state. When the grant
   * is approved it creates the token and consumes the grant in one transaction.
   * Success yields void; all other outcomes are tagged errors.
   */
  const pollDeviceAuthorization = (
    codeHash: string,
    secretHash: string,
    now: Date,
    token: APIToken,
  ): Effect.Effect<
    void,
    SqlError | GrantNotFound | GrantExpired | GrantDenied | GrantConsumed | GrantPending | SlowDown
  > =>
    sql.withTransaction(
      Effect.gen(function* () {
        const rows =
          yield* sql<DeviceRow>`${sql.unsafe(deviceSelect)} WHERE device_code_hash = ${codeHash} AND device_secret_hash = ${secretHash}`
        if (rows[0] === undefined) return yield* Effect.fail(new GrantNotFound())
        const g = toDevice(rows[0])
        if (g.expiresAt.getTime() <= now.getTime()) return yield* Effect.fail(new GrantExpired())
        if (g.status === "denied") return yield* Effect.fail(new GrantDenied())
        if (g.status === "consumed") return yield* Effect.fail(new GrantConsumed())
        if (g.status === "approved") {
          yield* sql`INSERT INTO api_tokens (id, token_hash, display_prefix, device_label, scopes, created_at, expires_at)
              VALUES (${token.id}, ${token.tokenHash}, ${token.displayPrefix}, ${token.deviceLabel}, ${token.scopes}, ${iso(token.createdAt)}, ${iso(token.expiresAt)})`
          yield* sql`UPDATE device_authorizations SET status = 'consumed', consumed_at = ${iso(now)} WHERE id = ${g.id} AND status = 'approved'`
          if ((yield* changes) !== 1) return yield* Effect.fail(new GrantConsumed())
          return
        }
        if (g.lastPollAt !== null && now.getTime() < g.lastPollAt.getTime() + g.pollIntervalSeconds * 1000) {
          const interval = Math.min(g.pollIntervalSeconds + 5, 30)
          yield* sql`UPDATE device_authorizations SET last_poll_at = ${iso(now)}, poll_interval_seconds = ${interval} WHERE id = ${g.id}`
          return yield* Effect.fail(new SlowDown({ interval }))
        }
        if (g.status === "pending") {
          yield* sql`UPDATE device_authorizations SET last_poll_at = ${iso(now)} WHERE id = ${g.id}`
          return yield* Effect.fail(new GrantPending({ interval: g.pollIntervalSeconds }))
        }
        return yield* Effect.fail(new GrantNotFound())
      }),
    )

  // --- api tokens ----------------------------------------------------------

  const tokenSelect = `SELECT id, token_hash, display_prefix, device_label, scopes, created_at, expires_at, last_used_at, revoked_at FROM api_tokens`

  const apiTokenByHash = (hash: string, now: Date): Effect.Effect<APIToken, SqlError | NotFound> =>
    Effect.flatMap(
      sql<TokenRow>`${sql.unsafe(tokenSelect)} WHERE token_hash = ${hash} AND revoked_at IS NULL AND expires_at > ${iso(now)}`,
      (rows) => (rows[0] === undefined ? Effect.fail(new NotFound()) : Effect.succeed(toToken(rows[0]))),
    )

  const touchAPIToken = (id: string, now: Date): Effect.Effect<void, SqlError> =>
    sql`UPDATE api_tokens SET last_used_at = ${iso(now)} WHERE id = ${id}
        AND (last_used_at IS NULL OR last_used_at <= ${iso(new Date(now.getTime() - 5 * 60 * 1000))})`.pipe(Effect.asVoid)

  const revokeAPIToken = (id: string, now: Date): Effect.Effect<void, SqlError | NotFound> =>
    sql.withTransaction(
      Effect.gen(function* () {
        yield* sql`UPDATE api_tokens SET revoked_at = ${iso(now)} WHERE id = ${id} AND revoked_at IS NULL`
        if ((yield* changes) === 0) return yield* Effect.fail(new NotFound())
      }),
    )

  const listAPITokens = (): Effect.Effect<ReadonlyArray<APIToken>, SqlError> =>
    Effect.map(sql<TokenRow>`${sql.unsafe(tokenSelect)} ORDER BY created_at DESC`, (rows) => rows.map(toToken))

  // --- admin sessions ------------------------------------------------------

  const createAdminSession = (session: AdminSession): Effect.Effect<void, SqlError | SessionLimit> =>
    sql.withTransaction(
      Effect.gen(function* () {
        const now = iso(session.createdAt)
        yield* sql`DELETE FROM admin_sessions WHERE expires_at <= ${now}`
        const active = (yield* sql<CountRow>`SELECT count(*) AS n FROM admin_sessions`)[0]!.n
        if (active >= maxAdminSessions) return yield* Effect.fail(new SessionLimit())
        yield* sql`INSERT INTO admin_sessions (id, session_hash, authenticated, created_at, expires_at)
            VALUES (${session.id}, ${session.sessionHash}, ${boolToInt(session.authenticated)}, ${now}, ${iso(session.expiresAt)})`
      }),
    )

  const adminSessionByHash = (hash: string, now: Date): Effect.Effect<AdminSession, SqlError | NotFound> =>
    Effect.flatMap(
      sql<SessionRow>`SELECT id, session_hash, authenticated, created_at, expires_at FROM admin_sessions
          WHERE session_hash = ${hash} AND expires_at > ${iso(now)}`,
      (rows) => (rows[0] === undefined ? Effect.fail(new NotFound()) : Effect.succeed(toSession(rows[0]))),
    )

  const authenticateAdminSession = (id: string, expiresAt: Date): Effect.Effect<void, SqlError> =>
    sql`UPDATE admin_sessions SET authenticated = 1, expires_at = ${iso(expiresAt)} WHERE id = ${id}`.pipe(Effect.asVoid)

  const deleteAdminSession = (id: string): Effect.Effect<void, SqlError> =>
    sql`DELETE FROM admin_sessions WHERE id = ${id}`.pipe(Effect.asVoid)

  return {
    create,
    get,
    save,
    del,
    list,
    deleteExpired,
    deleteExpiredAuth,
    createDeviceAuthorization,
    deviceAuthorizationByUserCode,
    deviceAuthorizationByDeviceCode,
    decideDeviceAuthorization,
    pollDeviceAuthorization,
    apiTokenByHash,
    touchAPIToken,
    revokeAPIToken,
    listAPITokens,
    createAdminSession,
    adminSessionByHash,
    authenticateAdminSession,
    deleteAdminSession,
  } as const
})

export class Store extends Context.Service<Store, Effect.Success<typeof make>>()("Store") {
  static readonly layer = Layer.effect(Store, make)
}
